### ISSUE TITLE
Javascript basics

### ISSUE THREAD

**Issue by SylarLong** [SylarLong/study-notes#4]
#### data type

Javascript provides two type of data: `basic type` and `referenced type`

##### basic type

- string
- number
- boolean
- null
- undefined
- Symbol
   Symbol is appended to Javascript from ES6.
   - values are different even two symbols defined with the same id
   ```js
   var s1 = Symbol('id');
   var s2 = Symbol('id');
   console.log(s1 === s2);  //false
   ```
   - it's un-iterator-able in object which means `for___in` and `Object.keys()` can __NOT__ access it
   ```js
   var s1 = Symbol('id');
   var obj = {
       [s1]: "a",
       b: "b"
   };

   console.log(Object.keys(s1));  // []

   for (var o in obj) {
       console.log(o);  // b
   }
   ```

##### referenced type

- Object
  - hasOwnProperty()
     sometimes we cannot make sure whether a property belongs to the specific object or not such as prototype properties.
  ```js

    function Human(name) {
        this.name = name;
        this.eyes = 2;
    }

    Human.prototype.greet = function () {
        console.log('hi there, I\'m', this.name);
    };

    function Asian(name, region) {
        Human.call(this, name);
        this.region = region;
    }

    Asian.prototype = Object.create(Human.prototype);
    Asian.prototype.constructor = Asian;

    var me = new Asian('Sylar', 'China');

    console.log('greet' in me);  // true
    console.log(!!me.greet);  // true
    console.log(me.hasOwnProperty('greet'));  // false

  ```
  - keys()
  - values()
  - entires()
  - assign()

- Array
  - from()
  - map()
  - forEach()
  - some()
  - find()
  - findIndex()
- Function

#### eval

security & performance issue.

#### prototype methods vs object methods
In short: prototype methods will always be the first choice. coz it's faster and memory efficient.

##### prototype methods
all instances will share prototype methods. so it costs less memory and easy to override.

##### object methods
it will be create for every single instance. it means it costs more memory and it's not possible to change the logic in common.
